알림 오류 해결

## 작업 요약
알림 오류 해결

### 작업 세부 사항
- [ ] 알림 신선도 관리 쿼리문 오류
```java
@Modifying
@Query("DELETE FROM Notification n WHERE (CURRENT_DATE - CAST(n.createdAt AS DATE)) >= 30")
void deleteAllIfAMonthOld();
```
CURRENT_DATE - CAST(n.createdAt AS DATE) 가 일수가 아닌 나노초로 계산되어 스케줄러 작동할때마다 모든 토큰이 사라지는 문제. 
- [ ] fcm 토큰 저장 api 요청 시, 이미 존재하는 토큰 저장 요청 시 아래와 같은 에러 발생
```text
2025-03-16 18:17:08.909 [http-nio-8080-exec-1] [ERROR] o.h.e.j.s.SqlExceptionHelper: ERROR: duplicate key value violates unique constraint "unique_token"
  Detail: Key (token)=([fcm 토큰]) already exists.
```

--- a/src/main/java/com/exchangediary/global/config/web/interceptor/JwtAuthenticationInterceptor.java
+++ b/src/main/java/com/exchangediary/global/config/web/interceptor/JwtAuthenticationInterceptor.java
@@ -12,6 +12,7 @@ import lombok.RequiredArgsConstructor;
 import org.springframework.web.servlet.HandlerInterceptor;
 
 import java.io.IOException;
+import java.util.Optional;
 
 @RequiredArgsConstructor
 public class JwtAuthenticationInterceptor implements HandlerInterceptor {
@@ -27,7 +28,10 @@ public class JwtAuthenticationInterceptor implements HandlerInterceptor {
     ) throws IOException {
         if (request.getRequestURI().equals("/login")) {
             try {
-                verifyJwtToken(request, response);
+                String token = verifyJwtToken(request, response);
+                Long memberId = jwtService.extractMemberId(token);
+                checkMemberExists(memberId);
+                request.setAttribute("memberId", memberId);
             } catch (UnauthorizedException exception) {
                 return true;
             }
@@ -36,7 +40,10 @@ public class JwtAuthenticationInterceptor implements HandlerInterceptor {
         }
 
         try {
-            verifyJwtToken(request, response);
+            String token = verifyJwtToken(request, response);
+            Long memberId = jwtService.extractMemberId(token);
+            checkMemberExists(memberId);
+            request.setAttribute("memberId", memberId);
         } catch (UnauthorizedException exception) {
             if (request.getRequestURI().startsWith("/api")) {
                 throw exception;
@@ -47,20 +54,18 @@ public class JwtAuthenticationInterceptor implements HandlerInterceptor {
         return true;
     }
 
-    private void verifyJwtToken(HttpServletRequest request,
+    private String verifyJwtToken(HttpServletRequest request,
                                 HttpServletResponse response
     ) throws UnauthorizedException {
         String token = getJwtTokenFromCookies(request);
-        String newToken = jwtService.verifyAccessToken(token);
+        Optional<Long> memberId = jwtService.verifyAccessToken(token);
 
-        if (newToken != null) {
+        if (memberId.isPresent()) {
+            String newToken = jwtService.generateAccessToken(memberId.get());
             cookieService.addCookie(jwtService.COOKIE_NAME, newToken, response);
-            token = newToken;
+            return newToken;
         }
-
-        Long memberId = jwtService.extractMemberId(token);
-        checkMemberExists(memberId);
-        request.setAttribute("memberId", memberId);
+        return token;
     }
 
     private String getJwtTokenFromCookies(HttpServletRequest request) {

--- a/src/main/java/com/exchangediary/member/service/AnonymousService.java
+++ b/src/main/java/com/exchangediary/member/service/AnonymousService.java
@@ -6,6 +6,8 @@ import jakarta.servlet.http.HttpServletResponse;
 import lombok.RequiredArgsConstructor;
 import org.springframework.stereotype.Service;
 
+import java.util.Optional;
+
 @Service
 @RequiredArgsConstructor
 public class AnonymousService {
@@ -14,20 +16,30 @@ public class AnonymousService {
     private final MemberQueryService memberQueryService;
 
     public AnonymousInfoResponse viewAnonymousInfo(String token, HttpServletResponse response) {
-        boolean shouldLogin = true;
+        token = needLogin(token, response);
+        boolean shouldLogin = (token == null);
         String groupId = null;
 
-        try {
-            token = jwtService.verifyAccessToken(token);
-
-            if (token != null) {
-                cookieService.addCookie(jwtService.COOKIE_NAME, token, response);
-            }
-            shouldLogin = false;
+        if (!shouldLogin) {
             Long memberId = jwtService.extractMemberId(token);
             groupId = memberQueryService.findGroupIdBelongTo(memberId).orElse(null);
-        } catch (UnauthorizedException ignored) {}
+        }
 
         return AnonymousInfoResponse.of(shouldLogin, groupId);
+    }
+
+    private String needLogin(String token, HttpServletResponse response) {
+        try {
+            Optional<Long> memberId = jwtService.verifyAccessToken(token);
+
+            if (memberId.isPresent()) {
+                String newToken = jwtService.generateAccessToken(memberId.get());
+                cookieService.addCookie(jwtService.COOKIE_NAME, newToken, response);
+                token = newToken;
+            }
+            return token;
+        } catch (UnauthorizedException exception) {
+            return null;
+        }
     }
 }

--- a/src/main/java/com/exchangediary/member/service/JwtService.java
+++ b/src/main/java/com/exchangediary/member/service/JwtService.java
@@ -5,7 +5,6 @@ import com.exchangediary.global.exception.serviceexception.UnauthorizedException
 import com.exchangediary.member.domain.entity.RefreshToken;
 import io.jsonwebtoken.Claims;
 import io.jsonwebtoken.ExpiredJwtException;
-import io.jsonwebtoken.Jwt;
 import io.jsonwebtoken.JwtException;
 import io.jsonwebtoken.Jwts;
 import io.jsonwebtoken.SignatureAlgorithm;
@@ -17,6 +16,7 @@ import org.springframework.stereotype.Service;
 
 import java.security.Key;
 import java.util.Date;
+import java.util.Optional;
 
 @Service
 @RequiredArgsConstructor
@@ -53,15 +53,15 @@ public class JwtService {
                 .compact();
     }
 
-    public String verifyAccessToken(String token) throws UnauthorizedException {
+    public Optional<Long> verifyAccessToken(String token) throws UnauthorizedException {
         try {
             verifyToken(token);
         } catch (ExpiredJwtException exception) {
             Long memberId = Long.valueOf(exception.getClaims().getSubject());
             verifyRefreshToken(memberId);
-            return generateAccessToken(memberId);
+            return Optional.of(memberId);
         }
-        return null;
+        return Optional.empty();
     }
 
     public Long extractMemberId(String token) {

--- a/src/main/java/com/exchangediary/notification/component/NotificationScheduler.java
+++ b/src/main/java/com/exchangediary/notification/component/NotificationScheduler.java
@@ -3,6 +3,7 @@ package com.exchangediary.notification.component;
 import com.exchangediary.notification.service.NotificationService;
 import com.exchangediary.notification.service.NotificationTokenService;
 import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
 import org.springframework.scheduling.annotation.Scheduled;
 import org.springframework.stereotype.Component;
 
@@ -11,6 +12,8 @@ import org.springframework.stereotype.Component;
 public class NotificationScheduler {
     private final NotificationService notificationService;
     private final NotificationTokenService notificationTokenService;
+    @Value("${fcm.token.expiration-day}")
+    private String fcmTokenExpirationDay;
 
     @Scheduled(cron = "0 0 9,21 * * *")
     public void pushWriteDiaryNotification() {
@@ -19,6 +22,6 @@ public class NotificationScheduler {
 
     @Scheduled(cron = "0 0 0 * * *")
     public void manageTokenFreshness() {
-        notificationTokenService.deleteOldTokens();
+        notificationTokenService.deleteOldTokens(Long.parseLong(fcmTokenExpirationDay));
     }
 }

--- a/src/main/java/com/exchangediary/notification/domain/NotificationRepository.java
+++ b/src/main/java/com/exchangediary/notification/domain/NotificationRepository.java
@@ -5,6 +5,7 @@ import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.Modifying;
 import org.springframework.data.jpa.repository.Query;
 
+import java.time.LocalDateTime;
 import java.util.List;
 
 public interface NotificationRepository extends JpaRepository<Notification, Long> {
@@ -22,6 +23,6 @@ public interface NotificationRepository extends JpaRepository<Notification, Long
             "WHERE m.orderInGroup = m.group.currentOrder AND d.id is NULL")
     List<String> findTokensNoDiaryToday();
     @Modifying
-    @Query("DELETE FROM Notification n WHERE CURRENT_DATE - CAST(n.createdAt AS DATE) >= 30")
-    void deleteAllIfAMonthOld();
+    @Query("DELETE FROM Notification n WHERE n.createdAt < :localDateTime")
+    void deleteAllByCreatedAtLessThan(LocalDateTime localDateTime);
 }

--- a/src/main/java/com/exchangediary/notification/domain/NotificationRepository.java
+++ b/src/main/java/com/exchangediary/notification/domain/NotificationRepository.java
@@ -25,4 +25,5 @@ public interface NotificationRepository extends JpaRepository<Notification, Long
     @Modifying
     @Query("DELETE FROM Notification n WHERE n.createdAt < :localDateTime")
     void deleteAllByCreatedAtLessThan(LocalDateTime localDateTime);
+    boolean existsByToken(String token);
 }

--- a/src/main/java/com/exchangediary/notification/domain/entity/Notification.java
+++ b/src/main/java/com/exchangediary/notification/domain/entity/Notification.java
@@ -10,6 +10,8 @@ import jakarta.persistence.GenerationType;
 import jakarta.persistence.Id;
 import jakarta.persistence.JoinColumn;
 import jakarta.persistence.ManyToOne;
+import jakarta.persistence.Table;
+import jakarta.persistence.UniqueConstraint;
 import jakarta.validation.constraints.NotNull;
 import lombok.AllArgsConstructor;
 import lombok.Builder;
@@ -24,6 +26,12 @@ import static lombok.AccessLevel.PROTECTED;
 @Builder
 @NoArgsConstructor(access = PROTECTED, force = true)
 @AllArgsConstructor(access = PRIVATE)
+@Table(name = "Notification", uniqueConstraints = {
+        @UniqueConstraint(
+                name = "unique_token",
+                columnNames = "token"
+        )
+})
 public class Notification extends BaseEntity {
     @Id
     @GeneratedValue(strategy = GenerationType.IDENTITY)

--- a/src/main/java/com/exchangediary/notification/service/NotificationTokenService.java
+++ b/src/main/java/com/exchangediary/notification/service/NotificationTokenService.java
@@ -66,7 +66,7 @@ public class NotificationTokenService {
     }
 
     @Transactional
-    public void deleteOldTokens() {
-        notificationRepository.deleteAllByCreatedAtLessThan(LocalDateTime.now().minusMonths(1));
+    public void deleteOldTokens(long expirationDay) {
+        notificationRepository.deleteAllByCreatedAtLessThan(LocalDateTime.now().minusDays(expirationDay));
     }
 }

--- a/src/main/java/com/exchangediary/notification/service/NotificationTokenService.java
+++ b/src/main/java/com/exchangediary/notification/service/NotificationTokenService.java
@@ -52,16 +52,17 @@ public class NotificationTokenService {
         return notificationRepository.findTokensNoDiaryToday();
     }
 
+    @Transactional
     public void saveNotificationToken(NotificationTokenRequest notificationTokenRequest, Long memberId) {
         Member member = memberQueryService.findMember(memberId);
-        Notification notification = Notification.builder()
-                .token(notificationTokenRequest.token())
-                .member(member)
-                .build();
+        boolean isDuplicated = notificationRepository.existsByToken(notificationTokenRequest.token());
 
-        try {
+        if (!isDuplicated) {
+            Notification notification = Notification.builder()
+                    .token(notificationTokenRequest.token())
+                    .member(member)
+                    .build();
             notificationRepository.save(notification);
-        } catch (DataIntegrityViolationException ignored) {
         }
     }
 

--- a/src/main/java/com/exchangediary/notification/service/NotificationTokenService.java
+++ b/src/main/java/com/exchangediary/notification/service/NotificationTokenService.java
@@ -10,6 +10,7 @@ import org.springframework.dao.DataIntegrityViolationException;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
 
+import java.time.LocalDateTime;
 import java.util.ArrayList;
 import java.util.List;
 
@@ -66,6 +67,6 @@ public class NotificationTokenService {
 
     @Transactional
     public void deleteOldTokens() {
-        notificationRepository.deleteAllIfAMonthOld();
+        notificationRepository.deleteAllByCreatedAtLessThan(LocalDateTime.now().minusMonths(1));
     }
 }

--- a/src/main/resources/application.properties
+++ b/src/main/resources/application.properties
@@ -27,12 +27,12 @@ spring.jpa.properties.hibernate.globally_quoted_identifiers=true
 kakao.client_id=${KAKAO_CLIENT_ID}
 kakao.redirect_uri=http://localhost:8080${KAKAO_REDIRECT_URI}
 
-# Jwt (access - 1hour, refresh - 30days)
+# Jwt (access - 1hour, refresh - 30days, milliseconds)
 security.jwt.secret-key=${JWT_SECRET_KEY}
 security.jwt.access-token.expiration-time=3600000
 security.jwt.refresh-token.expiration-time=259200000
 
-cookie.max-age.millisecond=604800000
+cookie.max-age.millisecond=259200000
 
 # thymeleaf
 spring.thymeleaf.prefix=classpath:/templates/
@@ -47,6 +47,7 @@ spring.servlet.multipart.max-request-size=10MB
 file.resources.location=${LOCAL_FILE_LOCATION}
 
 # fcm
+fcm.token.expiration-day=30
 fcm.api-key=${FCM_API_KEY}
 fcm.auth-domain=${FCM_AUTH_DOMAIN}
 fcm.project-id=${FCM_PROJECT_ID}

--- a/src/test/java/com/exchangediary/member/service/JwtServiceTest.java
+++ b/src/test/java/com/exchangediary/member/service/JwtServiceTest.java
@@ -5,6 +5,8 @@ import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
 import org.springframework.test.context.jdbc.Sql;
 
+import java.util.Optional;
+
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.springframework.test.context.jdbc.Sql.ExecutionPhase.BEFORE_TEST_METHOD;
 
@@ -29,8 +31,8 @@ public class JwtServiceTest {
         Long memberId = 1L;
         String token = jwtService.generateAccessToken(memberId);
 
-        String newToken = jwtService.verifyAccessToken(token);
+        Optional<Long> result = jwtService.verifyAccessToken(token);
 
-        assertThat(newToken).isNull();
+        assertThat(result.isPresent()).isFalse();
     }
 }

--- a/src/test/java/com/exchangediary/member/service/ReissueJwtTokenTest.java
+++ b/src/test/java/com/exchangediary/member/service/ReissueJwtTokenTest.java
@@ -10,6 +10,8 @@ import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
 import org.springframework.test.context.jdbc.Sql;
 
+import java.util.Optional;
+
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.springframework.test.context.jdbc.Sql.ExecutionPhase.BEFORE_TEST_METHOD;
 
@@ -35,9 +37,8 @@ public class ReissueJwtTokenTest {
         refreshTokenRepository.save(refreshToken);
 
         Thread.sleep(1000);
-        String newToken = jwtService.verifyAccessToken(token);
+        Optional<Long> memberId = jwtService.verifyAccessToken(token);
 
-        Long memberId = jwtService.extractMemberId(newToken);
-        assertThat(memberId).isEqualTo(member.getId());
+        assertThat(memberId.isPresent()).isTrue();
     }
 }

--- a/src/test/java/com/exchangediary/notification/api/NotificationApiTest.java
+++ b/src/test/java/com/exchangediary/notification/api/NotificationApiTest.java
@@ -35,7 +35,32 @@ public class NotificationApiTest extends ApiBaseTest {
     }
 
     @Test
-    void 알림_fcm_토큰_업데이트_성공() {
+    void 동일한_fcm_토큰_저장() {
+        RestAssured
+                .given().log().all()
+                .cookie("token", token)
+                .contentType(ContentType.JSON)
+                .body(new NotificationTokenRequest("token"))
+                .when().patch("/api/members/notifications/token")
+                .then().log().all()
+                .statusCode(HttpStatus.OK.value());
+
+        RestAssured
+                .given().log().all()
+                .cookie("token", token)
+                .contentType(ContentType.JSON)
+                .body(new NotificationTokenRequest("token"))
+                .when().patch("/api/members/notifications/token")
+                .then().log().all()
+                .statusCode(HttpStatus.OK.value());
+
+        List<Notification> notifications = notificationRepository.findByMemberId(member.getId());
+        assertThat(notifications).hasSize(1);
+        assertThat(notifications.get(0).getToken()).isEqualTo("token");
+    }
+
+    @Test
+    void 알림_fcm_토큰_여러개_저장() {
         RestAssured
                 .given()
                 .cookie("token", token)

--- a/src/test/resources/application.properties
+++ b/src/test/resources/application.properties
@@ -31,6 +31,7 @@ security.jwt.refresh-token.expiration-time=3600000
 cookie.max-age.millisecond=5400000
 
 # fcm
+fcm.token.expiration-day=30
 fcm.api-key=${FCM_API_KEY}
 fcm.auth-domain=${FCM_AUTH_DOMAIN}
 fcm.project-id=${FCM_PROJECT_ID}

